### PR TITLE
Unparse TableScan with projections, filters or fetch to SQL string

### LINKED ISSUE #12154: Unparse TableScan with pushdown projection
### Is your feature request related to a problem or challenge?

Consider the following case:
```rust
    let ctx = SessionContext::new();
    ctx.sql("create table t1 (a int, b int)").await?;
    ctx.sql("insert into t1 values (1, 2), (3, 4)").await?;

    let plan = ctx.sql(r#"select s1.a from (select s1.a, s1.b from (select a, b from t1) s1) s1"#).await?.into_optimized_plan()?;
    println!("optimized plan: {plan}");
    let unparse = plan_to_sql(&plan)?.to_string();
    println!("{}", format!("Unparsed SQL: {}", unparse));
```
The output result is
```
Optimized plan: SubqueryAlias: s1
  SubqueryAlias: s1
    TableScan: t1 projection=[a]
Unparsed SQL: SELECT * FROM t1 AS s1
```
The unparsing isn't equal to the original SQL. I think it should be
```
select s1.a from t1 as s1
```

### Describe the solution you'd like

Handle the pushdown projection at `Unparser:select_to_sql_recursively`
https://github.com/apache/datafusion/blob/23ccca9cd9bdf1791985f2b4abb6d5d8a4fc5ece/datafusion/sql/src/unparser/plan.rs#L240-L243

### Describe alternatives you've considered

_No response_

### Additional context

_No response_

### PR DESCRIPTION
**Pull request by goldmedal**
## Which issue does this PR close?

Closes #12154.

## Rationale for this change
Besides `projections`, I also handle the `filters` and `fetch` for TableScan.

## Are these changes tested?
yes

## Are there any user-facing changes?
no

### CODE REVIEW

File: datafusion/sql/src/unparser/plan.rs [hunk 1]
```diff
@@ -507,6 +523,72 @@ impl Unparser<'_> {
...
+        alias: Option<TableReference>,
+    ) -> Result<LogicalPlan> {
+        match plan {
+            LogicalPlan::TableScan(table_scan) => {
+                // TODO: support filters for table scan with alias
+                if alias.is_some() && !table_scan.filters().is_empty() {
+                    return not_impl_err!(
+                        "Subquery alias is not supported for table scan with pushdown filters"
+                    );
+                }
```
Comment (goldmedal) [inline datafusion/sql/src/unparser/plan.rs:537]: There are some issues with handling the table name prefix for the column. It isn't simple. I think we can create another issue for it. See the example in the unit tests.

**Review by devinjdangelo — APPROVED** [review]
Thanks @goldmedal! This looks reasonable to me. The only possible controversial change I notice is the added public methods in the expr crate. They seem reasonable enough to me though.

**Comment by goldmedal** [comment]
> Thanks @goldmedal! This looks reasonable to me. The only possible controversial change I notice is the added public methods in the expr crate. They seem reasonable enough to me though.

Thanks, @devinjdangelo. I removed the controversial parts to align with the coding style. I think they might not be necessary for this PR's purpose.

File: datafusion/expr/src/logical_plan/builder.rs [hunk 2]
```diff
@@ -333,6 +333,19 @@ impl LogicalPlanBuilder {
             .map(Self::new)
     }
 
+    /// Convert a table provider into a builder with a TableScan with filter and fetch
```
Comment (alamb) [inline datafusion/expr/src/logical_plan/builder.rs:336]: This seems like a reasonably helpful API in my opinion

File: datafusion/sql/tests/cases/plan_to_sql.rs [hunk 3]
```diff
@@ -607,6 +609,151 @@ fn sql_round_trip(query: &str, expect: &str) {
...
+        vec![col("id").gt(lit(1)), col("age").lt(lit(2))],
+    )?
+    .build()?;
+    let table_scan_with_two_filter = plan_to_sql(&table_scan_with_two_filter)?;
+    assert_eq!(
+        format!("{}", table_scan_with_two_filter),
+        "SELECT * FROM t1 WHERE ((t1.id > 1) AND (t1.age < 2))"
+    );
+
+    // TODO: support filters for table scan with alias
```
Comment (alamb) [inline datafusion/sql/tests/cases/plan_to_sql.rs:690]: I suggest we file a ticket for this gap and leave a link to it in this comment

**Review by alamb — APPROVED** [review]
Thanks @goldmedal  and @devinjdangelo  

I took a look at this PR and it looks good from my point of view. The PR has a conflict that needs to be resolved prior to merging.

cc @phillipleblanc  as I think you have been reviewing  code in this area as well

File: datafusion/sql/tests/cases/plan_to_sql.rs (hunk 3 above)
Comment (goldmedal) [reply datafusion/sql/tests/cases/plan_to_sql.rs:690]: filed https://github.com/apache/datafusion/issues/12368

File: datafusion/sql/src/unparser/plan.rs [hunk 4]
```diff
@@ -532,6 +548,73 @@ impl Unparser<'_> {
         }
     }
 
+    fn unparse_table_scan_pushdown(
+        plan: &LogicalPlan,
+        alias: Option<TableReference>,
+    ) -> Result<LogicalPlan> {
+        match plan {
+            LogicalPlan::TableScan(table_scan) => {
+                // TODO: support filters for table scan with alias.Remove this check after #12368 issue.
```
Comment (phillipleblanc) [inline datafusion/sql/src/unparser/plan.rs:557]:
```suggestion
                // TODO: support filters for table scan with alias. Remove this check after #12368 issue.
```

**Review by phillipleblanc — APPROVED** [review]
This PR makes sense to me. Thanks!

**Comment by alamb** [comment]
🚀  thanks @goldmedal  for the code and @devinjdangelo  and @phillipleblanc  for the review

**Comment by goldmedal** [comment]
Thanks @devinjdangelo @alamb @phillipleblanc for the review.

**Pull request merged**